File: Sorting algos/Quick_sort/quick_sort.cpp
//
// Created by nikhi on ३०-०४-२०१६.
//

#include <iostream>
#include <iomanip>
#include <cmath>
#include <cstdlib>
#include <algorithm>

using namespace std;

template <typename T> void quick_sort(T arr[],int start,int end){
   int pivot,temp,j,i;
    if(start<end){
        pivot = start;
        i = start;
        j = end;

        while(i<j){

            while(arr[i]<=arr[pivot] && i<end)i++;

            while(arr[j]>arr[pivot])j--;

            if(i<j) {
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        temp = arr[pivot];
        arr[pivot] = arr[j];
        arr[j]=temp;

        quick_sort(arr,start,j-1);
        quick_sort(arr,j+1,end);
    }
}

int main(){
    int i_arr[8] = {12,11,9,7,5,3,1,7};
    quick_sort(i_arr,0,7);

    for(int i=0;i<8;i++){
        cout<<i_arr[i]<<" ";
    }

    return 0;
}

File: Sorting algos/Quick_sort/QuickSort.py
def partition(a,start,end):
	pivot = a[start]
	left = start+1
	right = end
	done = False
	
	while not done:
		while left<=right and a[left] <= pivot:
			left = left + 1
		while left <= right and a[right] >= pivot:
			right = right-1
		
		if right<left:
			done = True
		else:
			temp = a[left]
			a[left] = a[right]
			a[right] = temp
	
	temp = a[start]
	a[start] = a[right]
	a[right] = temp
	return right 

def QuickSort(a,start,end):
	if(start<end):
		pivot = partition(a,start,end)
		QuickSort(a,start,pivot-1)
		QuickSort(a,pivot+1,end)
		return a

def main():
    a = [32,45,1,6,3,23,-2]
    a = QuickSort(a,0,len(a)-1)
    print(a)

if __name__=='__main__':
    main()
